app/src/main/java/com/islaharper/dawnofandroid/presentation/screens/login/LoginScreen.kt

```diff
@@ -47,47 +55,123 @@ fun LoginScreen(
                 signedInState = signedInState,
                 messageBarState = messageBarState,
                 onButtonClicked = {
-                    loginViewModel.saveSignedInState(signedIn = true)
-                },
+                    loginViewModel.oneTapSignIn()
+                }
             )
-        },
+        }
+    )
+
+    val launcher = rememberLauncherForActivityResult(
+        ActivityResultContracts.StartIntentSenderForResult()
+    ) { result ->
+        try {
+            if (result.resultCode == Activity.RESULT_OK) {
+                loginViewModel.verifyToken(intent = result.data)
+            } else {
+                loginViewModel.saveSignedInState(signedIn = false)
+            }
+        } catch (e: ApiException) {
+            loginViewModel.saveSignedInState(signedIn = false)
+        }
+    }
+
+    fun launch(signInResult: BeginSignInResult) {
+        val intent = IntentSenderRequest.Builder(signInResult.pendingIntent.intentSender).build()
+        launcher.launch(intent)
+    }
+
+    OneTapSignIn(
```

Comment:
Can you explain a bit more about this `OneTapSignIn` Composable? Is it intended to be a screen or something else (like a one-off that we trigger using `LaunchedEffect`)?

Reply:
Hi, this is used to display the one tap sign in UI:
https://developers.google.com/identity/one-tap/android/get-saved-credentials#3_display_the_one_tap_sign-in_ui

Reply:
This is going to take me some time to get around to investigating. In the meantime, is there any other work that can be made into small succinct PRs for quick approval? For example, PRs containing Composables with Previews that can be reviewed atomically?

Reply:
@DurtDiv3r I think we may be able to achieve a solution where this launching behavior is driven from the ViewModel. In general, we don't have to drive everything from the ViewModel (like the expanded/contracted state of a UI element that has no effect on business logic). But here it might help us understand better what is going on. 

Can we try starting with exposing state from the ViewModel that represents whether we should launch One Tap Sign In or not?

```kotlin
val launchOneTapSignIn: Boolean by loginViewModel.launchOneTapSignIn.collectAsState()

// snip

if (launchOneTapSignIn) {
   loginViewModel.onOneTapSignInLaunched() // sets state of boolean to false
   launcher.launch( // etc.)
}
```

The ViewModel should set this to `true` initially. Then after one tap sign in is launched, it should be set to `false` to prevent it from being retriggered. This Medium article from one of the Google devs explains it:

 https://medium.com/androiddevelopers/viewmodel-one-off-event-antipatterns-16a1da869b95 

would be good to take notes.

I think you would have to do the same for `navigateToHomeScreen`:

```kotlin
val navigateToHomeScreen: Boolean by loginViewModel.navigateToHomeScreen.collectAsState()

// snip

if (navigateToHomeScreen) {
   loginViewModel.onNavigateToHomeScreen() // set value to false
   navigator.navigate() // etc
}
```

When I looked at performing this refactor, one thing that is getting in the way is how `apiResponse` is being exposed by the ViewModel to the Composable. I think we should stop doing that. Concentrate on getting the ViewModel to make decisions by itself, and then tell the View about the result of the decisions so that the View can concentrate on what to display. 

Instead of this:

```kotlin
@Composable
fun OneTapSignIn(
    viewModel: LoginViewModel = hiltViewModel(),
    launch: (result: BeginSignInResult) -> Unit
) {
    when (val oneTapSignInResponse = viewModel.oneTapSignInResponse) {
        is Resource.Idle -> null
        is Resource.Error -> {
            viewModel.saveSignedInState(signedIn = false)
            viewModel.updateMessageBarErrorState(stringResource(R.string.google_account_not_found))
        }

        is Resource.Success -> oneTapSignInResponse.data.let {
            LaunchedEffect(key1 = it) {
                launch(it)
            }
        }
    }
}
```

It needs to be more like:
```kotlin
val showMessageBar: Int? by loginViewModel.showMessageBar.collectAsState()
```

Please let me know if this is unclear, we can spend some time working through this refactor together later in the week.